File: DesignPatterns/01_GOF23/DP_examples/Rust/01_Creational/Builder/concrete_builder.rs
use super::builder::{IBuilder, Item, ItemKind};
use super::product::{Burger, ColdDrink, VegetableBurger, ChickenBurger, CokeCola, PepsiCola};

// 汉堡建造者
pub struct BurgerBuilder<T: Burger + Default + 'static> {
    _marker: std::marker::PhantomData<T>,
}

impl<T: Burger + Default + 'static> BurgerBuilder<T> {
    pub const INSTANCE: BurgerBuilder<T> = BurgerBuilder {
        _marker: std::marker::PhantomData,
    };
}

impl<T: Burger + Default + 'static> Default for BurgerBuilder<T> {
    fn default() -> Self {
        Self::INSTANCE
    }
}

impl<T: Burger + Default + 'static> IBuilder for BurgerBuilder<T> {
    fn build(&self) -> Box<dyn Item> {
        Box::new(T::default())
    }
}

// 冷饮建造者
pub struct ColdDrinkBuilder<T: ColdDrink + Default + 'static> {
    _marker: std::marker::PhantomData<T>,
}

impl<T: ColdDrink + Default + 'static> ColdDrinkBuilder<T> {
    pub const INSTANCE: ColdDrinkBuilder<T> = ColdDrinkBuilder {
        _marker: std::marker::PhantomData,
    };
}

impl<T: ColdDrink + Default + 'static> Default for ColdDrinkBuilder<T> {
    fn default() -> Self {
        Self::INSTANCE
    }
}

impl<T: ColdDrink + Default + 'static> IBuilder for ColdDrinkBuilder<T> {
    fn build(&self) -> Box<dyn Item> {
        Box::new(T::default())
    }
}

// 套餐建造者接口
pub trait IMealBuilder {
    fn get_builders(&self) -> Vec<Box<dyn IBuilder>>;
}

// 鸡肉套餐建造者
pub struct ChickenMealBuilder;

impl ChickenMealBuilder {
    pub const INSTANCE: ChickenMealBuilder = ChickenMealBuilder;
}

impl IMealBuilder for ChickenMealBuilder {
    fn get_builders(&self) -> Vec<Box<dyn IBuilder>> {
        vec![
            Box::new(BurgerBuilder::<ChickenBurger>::INSTANCE),
            Box::new(ColdDrinkBuilder::<PepsiCola>::INSTANCE),
        ]
    }
}

impl Default for ChickenMealBuilder {
    fn default() -> Self {
        Self::INSTANCE
    }
}

// 素食套餐建造者
pub struct VegMealBuilder;

impl VegMealBuilder {
    pub const INSTANCE: VegMealBuilder = VegMealBuilder;
}

impl IMealBuilder for VegMealBuilder {
    fn get_builders(&self) -> Vec<Box<dyn IBuilder>> {
        vec![
            Box::new(BurgerBuilder::<VegetableBurger>::INSTANCE),
            Box::new(ColdDrinkBuilder::<CokeCola>::INSTANCE),
        ]
    }
}

impl Default for VegMealBuilder {
    fn default() -> Self {
        Self::INSTANCE
    }
}

// Meal Product
pub struct MealBuilder {
    builders: Vec<Box<dyn IBuilder>>,
    items: Vec<Box<dyn Item>>,
}

impl MealBuilder {
    pub fn new(builders: Vec<Box<dyn IBuilder>>) -> Self {
        Self {
            builders: builders,
            items: Vec::new(),
        }
    }
    
    pub fn add(&mut self, kind: ItemKind) {
        let builder: Box<dyn IBuilder> = match kind {
            ItemKind::VegetableBurger => Box::new(BurgerBuilder::<VegetableBurger>::INSTANCE),
            ItemKind::ChickenBurger => Box::new(BurgerBuilder::<ChickenBurger>::INSTANCE),
            ItemKind::CokeCola => Box::new(ColdDrinkBuilder::<CokeCola>::INSTANCE),
            ItemKind::PepsiCola => Box::new(ColdDrinkBuilder::<PepsiCola>::INSTANCE),
        };
        self.builders.push(builder);
    }
    
    pub fn remove(&mut self, kind: ItemKind) {
        let builder_to_remove: Box<dyn IBuilder> = match kind {
            ItemKind::VegetableBurger => Box::new(BurgerBuilder::<VegetableBurger>::INSTANCE),
            ItemKind::ChickenBurger => Box::new(BurgerBuilder::<ChickenBurger>::INSTANCE),
            ItemKind::CokeCola => Box::new(ColdDrinkBuilder::<CokeCola>::INSTANCE),
            ItemKind::PepsiCola => Box::new(ColdDrinkBuilder::<PepsiCola>::INSTANCE),
        };
        
        self.builders.retain(|b| !std::ptr::eq(b.as_ref() as *const dyn IBuilder, builder_to_remove.as_ref() as *const dyn IBuilder));
    }
    
    pub fn build(&mut self) {
        self.items.clear();
        for builder in &self.builders {
            self.items.push(builder.build());
        }
    }
    
    pub fn pack(&self) {
        for item in &self.items {
            item.packer().pack(item.as_ref());
        }
    }
    
    pub fn get_cost(&self) -> f32 {
        self.items.iter().map(|item| item.price()).sum()
    }
}

File: DesignPatterns/01_GOF23/DP_examples/Python/01_Creational/Builder/ConcreteBuilder.py
from Builder import IBuilder, IMealBuilder, Item
from Product import (
    Burger,
    ColdDrink,
    ItemKind,
    VegetableBurger,
    ChickenBurger,
    CokeCola,
    PepsiCola,
)
from typing import Type, List


# 汉堡建造者
class BurgerBuilder(IBuilder):
    def __init__(self, burger_class: Type[Burger]):
        self.burger_class = burger_class

    def Build(self):
        return self.burger_class()


# 冷饮建造者
class ColdDrinkBuilder(IBuilder):
    def __init__(self, drink_class: Type[ColdDrink]):
        self.drink_class = drink_class

    def Build(self):
        return self.drink_class()


# 鸡肉套餐建造者
class ChickenMealBuilder(IMealBuilder):
    _instance = None

    def GetInstance(self):
        if self._instance is None:
            self._instance = ChickenMealBuilder()
        return self._instance

    def GetBuilders(self) -> List[IBuilder]:
        return [BurgerBuilder(ChickenBurger), ColdDrinkBuilder(PepsiCola)]


# 素食套餐建造者
class VegMealBuilder(IMealBuilder):
    _instance = None

    def GetInstance(self):
        if self._instance is None:
            self._instance = VegMealBuilder()
        return self._instance

    def GetBuilders(self) -> List[IBuilder]:
        return [BurgerBuilder(VegetableBurger), ColdDrinkBuilder(CokeCola)]


class MealBuilder:
 

    def __init__(self, builders: List[IBuilder]):
        self._builders: List[IBuilder] = builders
        self._items: List[Item] = []

    def Add(self, kind: ItemKind):
        match kind:
            case ItemKind.VegetableBurger:
                self._builders.append(BurgerBuilder(VegetableBurger))
            case ItemKind.ChickenBurger:
                self._builders.append(BurgerBuilder(ChickenBurger))
            case ItemKind.CokeCola:
                self._builders.append(ColdDrinkBuilder(CokeCola))
            case ItemKind.PepsiCola:
                self._builders.append(ColdDrinkBuilder(PepsiCola))

    def Remove(self, kind: ItemKind):
        for builder in self._builders:
            if isinstance(builder, BurgerBuilder) and builder.burger_class.__name__ == kind.name:
                self._builders.remove(builder)
            elif isinstance(builder, ColdDrinkBuilder) and builder.drink_class.__name__ == kind.name:
                self._builders.remove(builder)

    def Build(self):
        for builder in self._builders:
            self._items.append(builder.Build())

    def Pack(self):
        for item in self._items:
            item.Packer.Pack(item)

    def GetCost(self) -> float:
        return sum(item.Price for item in self._items)
